File: go/classes/opaque.go
package classes

import "fmt"

type Fighter interface {
	init()
}

type XWing struct{}

func (x XWing) init() {}

type YWing struct{}

func (y YWing) init() {}

func lanchXFighter() Fighter {
	return XWing{}
}

func lanchYFighter() Fighter {
	return YWing{}
}

func compareFighter() {
	x := lanchXFighter()
	x1 := lanchXFighter()
	y := lanchYFighter()

	if x == x1 {
		fmt.Println("x == x1")
	}

	if x == y {
		fmt.Println("x == y")
	}
}

File: swift/Programming/Programming/Classes/Opaque.swift
//
//  Opaque.swift
//  Programming
//
//  Created by ke on 11/18/24.
//

protocol Fighter: Equatable {}
struct XWing: Fighter {}
struct YWing: Fighter {}

func lanchXFighter() -> some Fighter {
    return XWing()
}

func lanchYFighter() -> some Fighter {
    return YWing()
}


func compareFighter(){
    let x = lanchXFighter()
    let x2 = lanchXFighter()
    let y = lanchYFighter()
    
    if x == x2 {
        print("x == y")
    }
    
    //do not compile
//    if x == y {
//
//    }
}
